ArduPlane/mode_takeoff.cpp

```diff
@@ -68,66 +68,80 @@ bool ModeTakeoff::_enter()
 
 void ModeTakeoff::update()
 {
-    if (!takeoff_started) {
-        // see if we will skip takeoff as already flying
-        if (plane.is_flying() && (millis() - plane.started_flying_ms > 10000U) && ahrs.groundspeed() > 3) {
-            gcs().send_text(MAV_SEVERITY_INFO, "Takeoff skipped - circling");
-            plane.prev_WP_loc = plane.current_loc;
-            plane.next_WP_loc = plane.current_loc;
-            takeoff_started = true;
-            plane.set_flight_stage(AP_FixedWing::FlightStage::NORMAL);
-        }
-    }
-
-    if (!takeoff_started) {
-        // setup target location 1.5 times loiter radius from the
-        // takeoff point, at a height of TKOFF_ALT
-        const float dist = target_dist;
-        const float alt = target_alt;
-        const float direction = degrees(ahrs.yaw);
-
-        start_loc = plane.current_loc;
+    // don't setup waypoints if we dont have a valid position and home!
+    if (plane.current_loc.initialised() && AP::ahrs().home_is_set()) {
         plane.prev_WP_loc = plane.current_loc;
-        plane.next_WP_loc = plane.current_loc;
-        plane.next_WP_loc.alt += alt*100.0;
-        plane.next_WP_loc.offset_bearing(direction, dist);
-
-        plane.crash_state.is_crashed = false;
-
-        plane.auto_state.takeoff_pitch_cd = level_pitch * 100;
-
-        plane.set_flight_stage(AP_FixedWing::FlightStage::TAKEOFF);
-
-        if (!plane.throttle_suppressed) {
-            gcs().send_text(MAV_SEVERITY_INFO, "Takeoff to %.0fm at %.1fm to %.1f deg",
-                            alt, dist, direction);
-            takeoff_started = true;
-        }
-    }
-
-    // we finish the initial level takeoff if we climb past
-    // TKOFF_LVL_ALT or we pass the target location. The check for
-    // target location prevents us flying forever if we can't climb
-    if (plane.flight_stage == AP_FixedWing::FlightStage::TAKEOFF &&
-        (plane.current_loc.alt - start_loc.alt >= level_alt*100 ||
-         start_loc.get_distance(plane.current_loc) >= target_dist)) {
-        // reached level alt, re-calculate bearing to cope with systems with no compass
-        // or with poor initial compass
-        float direction = start_loc.get_bearing_to(plane.current_loc) * 0.01;
-        float dist_done = start_loc.get_distance(plane.current_loc);
         const float dist = target_dist;
+        const float alt = target_alt;  
+        if (!takeoff_started) {
+            const float direction = degrees(ahrs.yaw);
+            // if you are flying but had not used Mode Takeoff before
+            // then start_loc not set and use home as start location
+            start_loc = AP::ahrs().get_home();
```

Comment:
This bit should be in the "is flying" if statement.

Reply:
using home is a bit odd. I think you really want current location with home altitude? You we do a similar thing in QRTL to climb first, we use terrain alt there:

https://github.com/ArduPilot/ardupilot/blob/4a8eda058997548fd8979848f450b7e831274fa9/ArduPlane/mode_qrtl.cpp#L38-L48